### ISSUE TITLE
[QUERY] AzureJacksonAdapter fails to serialize map with key of resource ID

### ISSUE THREAD

**Issue by xseeseesee** [Azure/azure-sdk-for-java#8372]
**Query/Question**
We use AzureJacksonAdapter to serialize the [VirtualMachineInner](https://github.com/Azure/azure-libraries-for-java/blob/vnext-compute-fix-tests/azure-mgmt-compute/src/main/java/com/azure/management/compute/models/VirtualMachineInner.java). It doesn't work as expected when parsing the map inside the property of `identity`. 

***Why is this not a Bug or a feature Request?***
Not sure if we missed adding some specific annotation in the object.

Code snippet to reproduce:
```java
        VirtualMachineInner inner = new VirtualMachineInner();
        VirtualMachineIdentity identity = new VirtualMachineIdentity();
        Map<String, VirtualMachineIdentityUserAssignedIdentities> map = new HashMap<>();
        map.put("/subscriptions/0-0-0-0-0/resourcegroups/0/providers/Microsoft.ManagedIdentity/userAssignedIdentities/0",
                new VirtualMachineIdentityUserAssignedIdentities());
        identity.withType(ResourceIdentityType.SYSTEM_ASSIGNED_USER_ASSIGNED);
        identity.withUserAssignedIdentities(map);
        inner.withIdentity(identity);
```
If we only serialize identity, we will get what we need
```json
{
  "type": "SystemAssigned, UserAssigned",
  "userAssignedIdentities": {
    "/subscriptions/0-0-0-0-0/resourcegroups/0/providers/Microsoft.ManagedIdentity/userAssignedIdentities/0": {}
  }
}
```
But the actual what we get from inner is
```json
  "identity": {
    "type": "SystemAssigned, UserAssigned",
    "userAssignedIdentities": {
      "/subscriptions/0-0-0-0-0/resourcegroups/0/providers/Microsoft": {
        "ManagedIdentity/userAssignedIdentities/0": {}
      }
    }
  },
```

**Setup (please complete the following information if applicable):**
 - OS: Windows
 - IDE : IntelliJ
 - Version of the Library used
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [ ] Query Added
- [ ] Setup information Added

**Comment by xccc-msft** [comment]
It looks the serialization fails due to the existing logic to handle flattening properties as the resource ID will contain `.` in the format of `Microsoft.certainprovider`. I tried to use `@JsonInclude` but it doesn't help solve this. For now, my workaround is to add hard-coded checking block to filter out if the key contains `Microsoft.` in the `FlatteningSerializer`.

**Comment by weidongxu-microsoft** [comment]
core 1.3.0-beta1

Got one as well. `Map` get broken around the dot "Microsoft.ManagedIdentity".

```
      "userAssignedIdentities":{ 
         "/subscriptions/ec0aa5f7-9e78-40c9-85cd-535c6305b380/resourcegroups/javacsmrg320274676/providers/Microsoft":{ 
            "ManagedIdentity/userAssignedIdentities/msi-id4b759285":{ 

            },
            "ManagedIdentity/userAssignedIdentities/msi-id58966116":{ 

            }
```

Probably caused by 
https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningSerializer.java#L107

**Comment by joshfree** [comment]
@alzimmermsft can you please take a look?

**Comment by weidongxu-microsoft** [comment]
Class structure:
```
@JsonFlatten
public class SiteInner extends Resource {
    @JsonProperty(value = "identity")
    private ManagedServiceIdentity identity;

    @JsonProperty(value = "properties.state", access = JsonProperty.Access.WRITE_ONLY)
    private String state;
...
}

public final class ManagedServiceIdentity {
    @JsonProperty(value = "userAssignedIdentities")
    private Map<String, ManagedServiceIdentityUserAssignedIdentities> userAssignedIdentities;
...
}
```

SiteInner has `@JsonFlatten`.
Key of userAssignedIdentities in ManagedServiceIdentity got dot in it, and get broken.

**Comment by alzimmermsft** [comment]
Thank you for filing this issue @xccc-msft and investigating a potential root cause @weidongxu-microsoft.

Based on the conversation it appears that `JsonFlatten` will over eagerly split values on `.` leading to incorrect serialization. I will report back once I have a concrete solution to this problem.

**Comment by alzimmermsft** [comment]
Investigated with the help of @jianghaolu and we determined the reason that it is splitting the map key on `.` is due to `VirtualMachineIdentity` being `final` (https://github.com/Azure/azure-libraries-for-java/blob/vnext-compute-fix-tests/azure-mgmt-compute/src/main/java/com/azure/management/compute/VirtualMachineIdentity.java#L18), this triggers `FlatteningSerializer` to treat the `Map` property differently. It will early out on [this line](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningSerializer.java#L121) instead of hitting [this code path](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/util/serializer/FlatteningSerializer.java#L133).

A quick short term fix would be removing `final` from `VirtualMachineIdentity`.

**Comment by weidongxu-microsoft** [comment]
@alzimmermsft 
Thanks!
Sounds easy to mitigate. I will try it.

**Comment by alzimmermsft** [comment]
Hi @weidongxu-microsoft was this workaround able to resolve your issue?

**Comment by weidongxu-microsoft** [comment]
@alzimmermsft 
I take this is a mitigation that works. But it appears not to be a fix on root cause. Does it means we just stop generating any `final` class?

Or should we try plan some real fix on the root? The dot `.` in x-ms-client-flatten is actually causing more pain than this. We also observe that discriminator value with dot in it also got broken, e.g. `odata.type` used in monitor service. And `final` would not solve that. (OK, that would be taken as another issue.)

Even if there is no real fix to the root, could we at least let `JsonFlatten` to be customized to use a different token for flatten? Maybe a much more arcane UTF-8 char that has less chance to get conflict?
This should be easy to try out, and I can also help. But I would like we first agree to the general approach to this issue.

cc @joshfree @jianghaolu @yungezz

**Comment by weidongxu-microsoft** [comment]
Or if it need change on generator, it also fine.

Current generator will not add "final" to class designated as "JsonFlatten". However this bug happens to class that used as property of a "JsonFlatten" class.

**Comment by samvaity** [comment]
Fixed with ##22591 to allow flattening of static and final variables.
